src/main/java/org/springframework/data/keyvalue/core/PredicateQueryEngine.java

```diff
@@ -78,12 +79,13 @@ private List<?> sortAndFilterMatchingRange(Iterable<?> source, @Nullable Predica
 		return filterMatchingRange(tmp, criteria, offset, rows);
 	}
 
-	private static <S> List<S> filterMatchingRange(List<S> source, @Nullable Predicate criteria, long offset, int rows) {
+	@Contract("!null, _, _, _ -> !null")
```

Comment:
Hum, the return is never `null` and `source` cannot be `null`. You sure about this annotation?

Reply:
never null regardless of input - I don't mind dropping it.